More functions

percontation opened the pull request:
This branch adds some new functions (new to bsdnt; present in comparable bignum libraries):

* nn_get_str_raw, nn_set_str_raw – low-level non-decimal string conversion functions
* nn_popcount, nn_scan1, nn_hamdist_m – simple bit-level "measurement" functions.
* zz_powui – exponentiation by a constant.

wbhart commented:
I will probably have multiple comments on the code.

1) // should always be /* */ for a comment since the former is not supported by ISO C99.
2) Blocks should always start with the opening brace on the next line, just to keep the coding style consistent. This includes both for/if/while blocks and function blocks.
3) I'm not convinced zz_powui fits with the design of the library. I'd accept a zz_powi. Of course it should  assert that its argument is nonnegative. But there are no other ui functions, so I prefer not to add this one randomly.
4) In the test code, is it really necessary to put double parentheses around a <= test? It is "required" for = to indicate that you really mean = and not ==, but I'm not aware of such a requirement for <=.
5) There are a few spaces missing before and after + and - signs (we don't use them with * and / by the way) and after the parentheses on a cast.
6) Whether or not comments have a full stop or not should be standardised. Also, multiline comments should be in line with others in the source, e.g. comment starts on new line after /* or not.

Most of these are just conventions that are established elsewhere in the code, so may as well keep these consistent.

I will have another look at the code when these are dealt with. I want to keep this library as clean as I possibly can. I spent rather a lot of time making it as nice as possible.

wbhart commented:
You should probably also add your name to the list of contributors in the README, if it isn't already there.

Perhaps we should start an AUTHORS file.

Brian Gladman contributed the random generation functions, Antony Vennard wrote some of the build system. I wrote most of the rest.

I guess the others could just be listed as minor contributors without any bullet points. But please yourself. If you don't care much about attribution, adding your name to the README is fine also.

percontation commented:
I believe I addressed some of the concerns you raised. I'm not quite sure what you meant on 4 concerning double parentheses for a <= (I did do a double parentheses for a <<= assignment, but that's not in the test code). Can you point out which line you were referring to?

wbhart commented:
On 29 January 2018 at 22:35, Ryan Goulden <notifications@github.com> wrote:

> I believe I addressed some of the concerns you raised. I'm not quite sure
> what you meant on 4 concerning double parentheses for a <= (I did do a
> double parentheses for a <<= assignment, but that's not in the test code).
> Can you point out which line you were referring to?
>

You are right, I misread that. It's in zz.c and it's fine.

percontation commented:
> Does the powering test check powering of any multiprecision integers?

Sorry, not sure what you mean by this. I did the other things though.

Good call on small-base testing... I was in fact missing a bug on 2^i for i > 63. :|

wbhart commented:
On 30 January 2018 at 23:33, Ryan Goulden <notifications@github.com> wrote:

> Does the powering test check powering of any multiprecision integers?
>
> Sorry, not sure what you mean by this.
>

I mean, does the test of nn_powi do any tests of a^b where a is bigger than
64 bits, i.e. is multiple words?

percontation commented:
Ah. Yes, the first test calls `randoms_signed` to generate a base `a` of up to 10 words. (Unless I'm confused about how the test functions work.)

wbhart commented:
Ah ok. I had forgotten too.

Then I think I can merge this now.
